component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -199,4 +212,119 @@ private List<String> populateMandatoryClaims() {
+     * @throws OAuthClientAuthnException
+     */
+    private String getRegisteredSigningAlgorithm(String clientId) throws OAuthClientAuthnException {
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (!(isNotEmpty(serviceProvider.getCertificateContent()))) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (Constants.TOKEN_ENDPOINT_AUTH_SIGNING_ALG.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+        return Constants.NOT_APPLICABLE;
+    }
+
+    /**
+     * Obtain the request signing algorithm from the signed JWT.
+     *
+     * @param signedObject   Signed JWT in the request.
+     * @return Request signing algorithm.
+     * @throws OAuthClientAuthnException
+     */
+    private String getRequestSigningAlgorithm(String signedObject) throws OAuthClientAuthnException {
+        try {
+            SignedJWT signedJWT = SignedJWT.parse(signedObject);
+            return signedJWT.getHeader().getAlgorithm().getName();
+        } catch (ParseException e) {
+            throw new OAuthClientAuthnException("Error occurred while parsing the signed assertion",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+    }
+
+    /**
+     * Check whether the application should support FAPI.
+     *
+     * @param clientId       Client ID of the application.
+     * @return Whether the application should support FAPI.
+     * @throws OAuthClientAuthnException
+     */
+    private String isFapiApplication(String clientId) throws OAuthClientAuthnException {
+
+        String isFapiApp = "IsFAPIApp";
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (isEmpty(serviceProvider.getCertificateContent())) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (isFapiApp.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Error occurred while retrieving the service provider",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+        return null;
```

Comment:
Returning null from a method is not a good practice. If we have to return null we have to handle it in all the references. This method reference does not handle null. https://github.com/wso2-extensions/identity-oauth-addons/pull/108/files#diff-09ec4bf87542925a5d149e923a23729cd910941e83b51658f6c66563c4055fe8R116 There is a possible Nullpointer.

Reply:
Fixed in c8efc7a5d86d1df217e9bdcab4167db2d8a20ab5